python/cuml/ensemble/randomforest_shared.pyx

```diff
@@ -0,0 +1,131 @@
+from cpython.object cimport PyObject
+from libc.stdint cimport uintptr_t
+from typing import Tuple, Dict, List, Union
+import numpy as np
+
+cdef extern from "treelite/tree.h" namespace "treelite":
+    cdef struct PyBufferFrame:
+        void* buf
+        char* format
+        size_t itemsize
+        size_t nitem
+    cdef cppclass Model:
+        vector[PyBufferFrame] GetPyBuffer() except +
+        void InitFromPyBuffer(vector[PyBufferFrame] frames) except +
+
+cdef extern from "Python.h":
+    Py_buffer* PyMemoryView_GET_BUFFER(PyObject* mview)
+
+cdef class PyBufferFrameWrapper:
+    cdef PyBufferFrame _handle
+    cdef Py_ssize_t shape[1]
+    cdef Py_ssize_t strides[1]
+
+    def __cinit__(self):
+        pass
+
+    def __dealloc__(self):
+        pass
+
+    def __getbuffer__(self, Py_buffer* buffer, int flags):
+        cdef Py_ssize_t itemsize = self._handle.itemsize
+
+        self.shape[0] = self._handle.nitem
+        self.strides[0] = itemsize
+
+        buffer.buf = self._handle.buf
+        buffer.format = self._handle.format
+        buffer.internal = NULL
+        buffer.itemsize = itemsize
+        buffer.len = self._handle.nitem * itemsize
+        buffer.ndim = 1
+        buffer.obj = self
+        buffer.readonly = 0
+        buffer.shape = self.shape
+        buffer.strides = self.strides
+        buffer.suboffsets = NULL
+
+    def __releasebuffer__(self, Py_buffer *buffer):
+        pass
+
+cdef PyBufferFrameWrapper MakePyBufferFrameWrapper(PyBufferFrame handle):
+    cdef PyBufferFrameWrapper wrapper = PyBufferFrameWrapper()
+    wrapper._handle = handle
+    return wrapper
+
+cdef list _get_frames(ModelHandle model):
+    frames = []
+    cdef Model* model_obj = <Model*>model
+    cdef vector[PyBufferFrame] interface = model_obj.GetPyBuffer()
+    cdef vector[PyBufferFrame].iterator it = interface.begin()
+    while it != interface.end():
+        v = deref(it)
+        w = MakePyBufferFrameWrapper(v)
+        frames.append(memoryview(w))
+        inc(it)
+    return frames
+
+cdef ModelHandle _init_from_frames(vector[PyBufferFrame] frames) except *:
+    cdef Model* model_obj = new Model()
+    model_obj.InitFromPyBuffer(frames)
+    cdef ModelHandle handle = <ModelHandle>model_obj
+    return handle
+
+
+def get_frames(model: uintptr_t) -> List[memoryview]:
+    return _get_frames(<ModelHandle> model)
+
+
+def init_from_frames(frames: List[np.ndarray],
+                     format_str: List[str], itemsize: List[int]) -> uintptr_t:
+    cdef vector[PyBufferFrame] cpp_frames
+    cdef Py_buffer* buf
+    cdef PyBufferFrame cpp_frame
+    for i, frame in enumerate(frames):
+        x = memoryview(frame)
+        buf = PyMemoryView_GET_BUFFER(<PyObject*>x)
+        cpp_frame.buf = buf.buf
+        cpp_frame.format = format_str[i]
+        cpp_frame.itemsize = itemsize[i]
+        cpp_frame.nitem = buf.len // itemsize[i]
+        cpp_frames.push_back(cpp_frame)
```

Comment:
Unfortunately, `.view()` is not suitable here, because NumPy interprets the format string `=l` as 8 bytes, even though the `=` prefix should have indicated the use of the "standard size" for `long`. (The standard size for `long` is 4 bytes.)

Reply:
Alternatively we can use [`memoryview.cast(...)`]( https://docs.python.org/3/library/stdtypes.html#memoryview.cast ), which doesn't have this issue.

Reply:
`memory.cast()` doesn't work either. Quote:

> The destination format is restricted to a single element native format in struct syntax.

The format string `format_str[i]` is more than one single letter. In particular, the frames corresponding to tree nodes have format string `T{=l=l=L=f=Q=d=d=b=b=?=?=?=?=H}`.

Reply:
Oh interesting. Didn't realize that. Yeah let's skip this then. Thanks for the context 🙂